### PR TITLE
feature/aql distinct collect executor

### PR DESCRIPTION
**Pull request by hkernbach**

### CODE REVIEW

File: arangod/Aql/CollectNode.cpp [hunk 1]
```diff
@@ -126,8 +128,42 @@ std::unique_ptr<ExecutionBlock> CollectNode::createBlock(
...
+      transaction::Methods* trxPtr = _plan->getAst()->query()->trx();
+      DistinctCollectExecutorInfos infos(getRegisterPlan()->nrRegs[previousNode->getDepth()],
+                                         getRegisterPlan()->nrRegs[getDepth()],
+                                         getRegsToClear(), std::move(readableInputRegisters),
+                                         std::move(writeableInputRegisters),
+                                         std::move(groupRegisters), trxPtr);
+
+      return std::make_unique<ExecutionBlockImpl<DistinctCollectExecutor>>(&engine, this,
+                                                                           std::move(infos));
+    }
```
Comment (goedderz) [inline arangod/Aql/CollectNode.cpp]: As we replace the other collect blocks as well, this should probably be unified as possible, extracted into separate functions, or both. Don't do it now, we will have a better idea how to do it when more blocks are done. Just wanted to mention it early.

File: arangod/Aql/DistinctCollectExecutor.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+  /// @brief input/output variables for the collection (out, in)
+  std::vector<std::pair<Variable const*, Variable const*>> _groupVariables;
+
+  /// @brief the transaction for this query
+  transaction::Methods* _trxPtr;
+};
+
+/**
+ * @brief Implementation of Limit Node
```
Comment (goedderz) [inline arangod/Aql/DistinctCollectExecutor.h]: copy&paste?

File: arangod/Aql/DistinctCollectExecutor.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+ */
+
+class DistinctCollectExecutor {
+ public:
+  struct Properties {
+    static const bool preservesOrder = true;
```
Comment (goedderz) [inline arangod/Aql/DistinctCollectExecutor.h]: This has to be false, this block does not preserve the order of the input rows.

**Comment by hkernbach** [comment]
http://jenkins01:8080/job/arangodb-matrix-pr-linux/4277/ (green)

**Comment by hkernbach** [comment]
> No tests? 😢

Sorry, tests are coming soon :)

File: arangod/Aql/DistinctCollectExecutor.h (hunk 3 above)
Comment (hkernbach) [reply arangod/Aql/DistinctCollectExecutor.h]: True

File: arangod/Aql/DistinctCollectExecutor.h (hunk 2 above)
Comment (hkernbach) [reply arangod/Aql/DistinctCollectExecutor.h]: Fixed

**Comment by hkernbach** [comment]
To be sure: 
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/3039/

File: tests/Aql/DistinctCollectExecutorTest.cpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        AND_THEN("the executor should return DONE") {
+          std::tie(state, stats) = testee.produceRow(result);
+          REQUIRE(state == ExecutionState::DONE);
+          REQUIRE(!result.produced());
+        }
+      }
+    }
+  }
+}
+
```
Comment (mchacki) [inline tests/Aql/DistinctCollectExecutorTest.cpp]: Can we add tests with data? Please.

Comment (goedderz) [reply tests/Aql/DistinctCollectExecutorTest.cpp]: +1

**Pull request closed without merging**